### DISCUSSION TITLE
disallow_empty, disallow_unknown, and None values

### DISCUSSION THREAD

**Discussion by topherbuckley** [eprbell/dali-rp2#178] [Q&A]
Can someone help me understand the intent behind `disallow_empty`? I thought I understood, that it would check if if some `value` had `len(value)==0` as per [here](https://github.com/eprbell/dali-rp2/blob/4969740753cc453810e7d384d8534b4d177f752a/src/dali/abstract_transaction.py#L54) but I'm quite confused by things like 

```python
    def _validate_optional_string_field(cls, name: str, value: Optional[str], raw_data: str, disallow_empty: bool, disallow_unknown: bool) -> Optional[str]:
        if not value:
            return None
        return cls._validate_string_field(name, value, raw_data, disallow_empty, disallow_unknown)
```
from [here](https://github.com/eprbell/dali-rp2/blob/4969740753cc453810e7d384d8534b4d177f752a/src/dali/abstract_transaction.py#L90-L93)

It would seem the check for whether something was `len(value)==0` will never be checked because the above statement will return `None` instead of error out when `value==None`. I would have thought `None` is also "empty", but it seems this is interpreted differently here. 

The main reason I ask is because I was trying to interpret the "required" and "optional" parameters to each of the In/Out/Intra transactions listed in the [config readme](https://github.com/eprbell/dali-rp2/blob/main/docs/configuration_file.md#manual-section-csv) with regards to these `disallow_empty` and `disallow_unknown`. Generally speaking it seems that `disallow_unknown` appears to be of no consequence to being "required" or "optional", whereas anything that is `disallow_empty` would be required, and anything that is allowed to be empty, is optional. The only outlier there is the `fiat_ticker` which appears to be "optional" due to the fact that this value is filled in with the native fiat in a few places in the [transaction_resolver](https://github.com/eprbell/dali-rp2/blob/4969740753cc453810e7d384d8534b4d177f752a/src/dali/transaction_resolver.py#L239)

```python
if transaction.fiat_ticker not in {None, global_configuration[Keyword.NATIVE_FIAT.value]}:
```

This seems very opaque to me, since [here](https://github.com/eprbell/dali-rp2/blob/4969740753cc453810e7d384d8534b4d177f752a/src/dali/abstract_transaction.py#L127-L129)
```python
        self.__fiat_ticker: Optional[str] = self._validate_optional_string_field(
            "fiat_ticker", fiat_ticker, raw_data, disallow_empty=True, disallow_unknown=True
        )
```
It is `disallow_empty`, so I would think that a `None` value would fail. These validate methods appear to let these `None` values through, and some are handled by transaction_resolver. 

So in the end, I am planning to just add my own field specifying `is_required` in some [AbstractTransactionItem class](https://github.com/eprbell/dali-rp2/discussions/176) to help clear these things up, but if there is a way to determine if something is required or not just by the values of `disallow_empty` and `disallow_unknown` please let me know. 

Looking at it all again, I guess the easiest way for me to determine if something is required or optional is to just look at the method call where it is verified (e.g. _validate_optional_string_field vs _validate_string_field). Is there a reason it was implemented this way? In other words, I'm curious if it doesn't make more things more clear to have a third `bool` called `required` in all of the constructors for things extending AbstractTransaction, and then just have one unified validate method for strings. I guess you lose some type hinting for the return value then. Anyways, I think I can add some clarity on this after getting my AbstractTransactionItem class finished.

**Comment by eprbell** [comment]
Regarding `_validate_optional_string_field` I think what you found a bug. The proper check should be:
```
if value is None:
   ...
```

Thanks for digging through this code and finding bugs.

As for the other question on optional vs required, the easy way to tell if a parameter is optional or required is just to look at the signature of the constructor: if a parameter has `Optional` in the type hint it's optional, otherwise it's required.

However I think I am missing some context: can you elaborate on why you need optional/required information in your PR and how you plan to use it?

**Reply by topherbuckley** [reply]
>The proper check should be:
>```
>if value is None:
>   ...
>```

Ok, so the `None` != "empty" is intentional, but as it stands they were not differentiated because of the more broad "falsy" check? 

>As for the other question on optional vs required, the easy way to tell if a parameter is optional or required is just to look at the signature of the constructor: if a parameter has Optional in the type hint it's optional, otherwise it's required.

Thanks that helps! 

>However I think I am missing some context: can you elaborate on why you need optional/required information in your PR and how you plan to use it?

I already manually added everything I needed, so I can elaborate on this within the [PR](https://github.com/eprbell/dali-rp2/pull/134). I'll try to update it to a point of working tests sometime tomorrow.

**Reply by topherbuckley** [reply]
>However I think I am missing some context: can you elaborate on why you need optional/required information in your PR and how you plan to use it?

I just wanted to add some more context to this as I'm still struggling to get things the way I want. 

I'm trying to understand when a value should be replaced with "__unknown" and when I should let it error out somewhere else down the stack. 

For example [here](https://github.com/eprbell/dali-rp2/blob/4969740753cc453810e7d384d8534b4d177f752a/src/dali/plugin/input/csv/manual.py#L231-L245):

```python
                    IntraTransaction(
                        plugin=self.__MANUAL,
                        unique_id=line[self.__INTRA_UNIQUE_ID_INDEX],
                        raw_data=raw_data,
                        timestamp=line[self.__INTRA_TIMESTAMP_INDEX],
                        asset=line[self.__INTRA_ASSET_INDEX],
                        from_exchange=from_exchange if from_exchange else Keyword.UNKNOWN.value,
                        from_holder=from_holder if from_holder else Keyword.UNKNOWN.value,
                        to_exchange=to_exchange if to_exchange else Keyword.UNKNOWN.value,
                        to_holder=to_holder if to_holder else Keyword.UNKNOWN.value,
                        spot_price=line[self.__INTRA_SPOT_PRICE_INDEX],
                        crypto_sent=crypto_sent if crypto_sent else Keyword.UNKNOWN.value,
                        crypto_received=crypto_received if crypto_received else Keyword.UNKNOWN.value,
                        notes=line[self.__INTRA_NOTES_INDEX],
                    )
```
So although both `asset` and `from_exchange` are "required", we substitute `__unknown` for the latter, but not the former. Why is this or how can I determine in general when one input_parameter should be:
1. left alone to be error handled down the stack
2. replaced with `__unknown`?

As it stands, (2.) is done if the values are either None, or "empty". As an example, for `from_exchange` above, even though it is a "required" value, it is acceptable to have the value as "empty" (e.g '') in the input csv file. This seems very counterintuitive to me.

**Reply by topherbuckley** [reply]
@macanudo527 just tagging you into this as I know you worked on the manual csv input plugin, so curious if you have any rational you can share with the above comment.

**Reply by macanudo527** [reply]
@topherbuckley Sorry, I wish I could help you out, but I didn't work on `manual.py` this is part of the core that @eprbell developed.

**Reply by eprbell** [reply]
> @macanudo527 just tagging you into this as I know you worked on the manual csv input plugin, so curious if you have any rational you can share with the above comment.

The code you pasted handles intra transactions in the manual plugin. Such transactions can be "partial", which is why holder and exchange fields can be left blank: take a look at the relevant section of the [manual plugin documentation](https://github.com/eprbell/dali-rp2/blob/main/docs/configuration_file.md#partial-transactions-and-transaction-resolution), which explains the semantics. For example you may have a transaction from Coinbase (which is supported by RP2) to an exchange that is not supported. In this case you let the Coinbase plugin handle the send side of the transaction and then you add the receiving half of the transaction in the manual csv file. The manual csv transaction would have only to_exchange and to_holder populated (from_exchange and from_holder are left empty and they are assigned __unknown by the code). Hope this helps.

**Reply by topherbuckley** [reply]
@eprbell thanks! Yes, I'm sure that will set me off in the right direction and gives me context with which to logically figure out why things are one way over another. I'll come back to this a few days, but will remind myself to look down the stack to see how these `from_exchange` and `from_holder` are error checked later on (i.e. why `__unknown` would be more appropriate than None or Empty). I think this should help clarify a few things.

**Reply by topherbuckley** [reply]
> @topherbuckley Sorry, I wish I could help you out, but I didn't work on `manual.py` this is part of the core that @eprbell developed.

Whoops, sorry, my mistake. Not sure where I came up with that.
